docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+current node usage.
+
+At each monitoring interval, if a compute resource has reached it's eviction
+threshold, the `kubelet` will record an event related to the node describing
+which resource is low.
+
+If a compute resource has reached its eviction threshold over the `30s` period
+the `kubelet` will initiate eviction to attempt to reclaim the resource that
+has met its eviction threshold.  The `kubelet` will record an event when eviction
+is initiated on the node.
+
+## Eviction of Pods
+
+If the eviction criteria have been met, the `kubelet` will evict containers
+based on their quality of service guarantee.  Within a particular quality of
+service tier, the `kubelet` may use an eviction strategy to determine the
+candidate set of pods to evict.
+
+The `kubelet` will accept a flag to define the eviction strategy as follows:
+
+* `--eviction-strategy=:"` - The strategy to use when evicting pods
+when eviction criteria have been met.
+
+The `kubelet` will target `BestEffort`, `Burstable`, and `Guaranteed`
+pods in that order to find a candidate set of pods to fail to force reclamation
+of the starved compute resource.  Within each quality of service tier,
+the `kubelet` will evaluate each pod relative to the eviction strategy.
+
+If the `kubelet` fails a pod based on eviction criteria, an `Event` is created
+describing why the `Pod` was failed by the `kubelet`.
+
+This proposal describes the `greedy` strategy as the first candidate to
+implement based on the ability for operators to easily reason about it.
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
+starved compute resource relative to their scheduling guarantee.
+
+It ranks pods within a quality of service tier as follows:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+### Other potential eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
+
+For example, operators may want to target `BestEffort` pods that were most
+recently scheduled on the node instead of targeting `BestEffort` pods that
+consumed the most resource.
+
+Another potential option is that operators may want to target `BestEffort` pods
+that consume the least amount of resource assuming their sum would be enough
+to reclaim enough resource on the node.  The idea being that those low resource
+pods could fit elsewhere on the cluster.
+
+Finally, operators may want to target pods based on other criteria altogether.
```

Comment:
Or data locality.. If the pod has to run on the same node, its better to not evict it.
Example DB's that use local SSDs.